generateDotGraph has re-entrance problem resulting in node ID conflict

I encounter a problem using gorgonia to generate the dot graph.

When I called twice in a same process (or go test to run different testing functions including dot.Marshal(g) in more than two places. The execution will produce
```
Error:
panic: simple: node ID collision: 3 [recovered]
	panic: simple: node ID collision: 3

goroutine 21 [running]:
testing.tRunner.func1.1(0x190dd20, 0xc00027e730)
	/usr/local/go/src/testing/testing.go:941 +0x3d0
testing.tRunner.func1(0xc0000cde60)
	/usr/local/go/src/testing/testing.go:944 +0x3f9
panic(0x190dd20, 0xc00027e730)
	/usr/local/go/src/runtime/panic.go:967 +0x166
gonum.org/v1/gonum/graph/simple.(*DirectedGraph).AddNode(0xc000212e10, 0x1b1a180, 0xc0000a2d70)
	/Users/wzzhu/go/pkg/mod/gonum.org/v1/gonum@v0.0.0-20191009222026-5d5638e6749a/graph/simple/directed.go:49 +0x206
gorgonia.org/gorgonia/encoding/dot.generateDotGraph(0xc0000f9860, 0x16299b6, 0xc0000f5500, 0x1b2e580, 0xc000250b40)
	/Users/wzzhu/go/pkg/mod/gorgonia.org/gorgonia@v0.9.3/encoding/dot/topgraph.go:49 +0x320
```
The conflict ID is random.

I found probably it was due to the reentrance problem in generateDotGraph (topgraph.go). subGraph is a global variable having side effects in a same process.
```
var (
	subGraphs = map[internalEncoding.Group]subgrapher{
		internalEncoding.ConstantCluster: constantSubGraph{
			DirectedBuilder: simple.NewDirectedGraph(),
			name:            "Constants",
		},
		internalEncoding.InputCluster: inputSubGraph{
			DirectedBuilder: simple.NewDirectedGraph(),
			name:            "Inputs",
		},
		internalEncoding.ExprGraphCluster: exprSubGraph{
			DirectedBuilder: simple.NewDirectedGraph(),
			name:            "ExprGraph",
			subs:            make(map[internalEncoding.Group]operatorSubGraph),
		},
		internalEncoding.UndefinedCluster: exprSubGraph{
			DirectedBuilder: simple.NewDirectedGraph(),
			name:            "ExprGraph",
		},
	}
)

type attributer []encoding.Attribute

func (a attributer) Attributes() []encoding.Attribute { return a }

func generateDotGraph(g *gorgonia.ExprGraph) (graph.Graph, error) {
...
				if subgrapher, ok := subGraphs[group]; ok {
					subgrapher.(graph.DirectedBuilder).AddNode(n) 

```
A simple test snippets like this will reproduce it.
```
func TestDupDotCalls(t *testing.T) {
	g := gorgonia.NewGraph()
	var x, y, z *gorgonia.Node
	var err error

	// define the expression
	x = gorgonia.NewScalar(g, gorgonia.Float64, gorgonia.WithName("x"))
	y = gorgonia.NewScalar(g, gorgonia.Float64, gorgonia.WithName("y"))
	if z, err = gorgonia.Add(x, y); err != nil {
		t.Fatal(err)
	}
	b, err := dot.Marshal(g)
	if err != nil {
		glog.Errorf("Error printing graph: %v", err)
	}
	fmt.Printf("\n%s", string(b))

	var x2, y2, z2 *gorgonia.Node

	g2 := gorgonia.NewGraph()
	x2 = gorgonia.NewScalar(g2, gorgonia.Float64, gorgonia.WithName("x2"))
	y2 = gorgonia.NewScalar(g2, gorgonia.Float64, gorgonia.WithName("y2"))
	if z2, err = gorgonia.Add(x2, y2); err != nil {
		t.Fatal(err)
	}
	b2, err := dot.Marshal(g2)
	if err != nil {
		t.Errorf("Error printing graph: %v", err)
	}
	fmt.Printf("\n%s", string(b2))
	t.Logf("%v %v", z, z2)
}


```

--- a/encoding/dot/issues_test.go
+++ b/encoding/dot/issues_test.go
@@ -5,6 +5,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
 	"gorgonia.org/gorgonia"
 )
 
@@ -28,7 +29,7 @@ func unmangleName(marshalled string, node *gorgonia.Node) string {
 func TestIssue_407(t *testing.T) {
 	// originally written by @wzzhu
 
-	//assert := assert.New(t)
+	assert := assert.New(t)
 	g := gorgonia.NewGraph()
 	var x, y, z *gorgonia.Node
 	var err error
@@ -64,7 +65,16 @@ func TestIssue_407(t *testing.T) {
 	snd = unmangleName(snd, x2)
 	snd = unmangleName(snd, y2)
 	snd = unmangleName(snd, z2)
-	//assert.Equal(fst, snd, "XXX")
-	// they are the same but due to Gonum's marshaling issue, it is marshalled undeterministically using a map.
+	assert.Equal(fst, snd, "XXX")
 	t.Logf("%v %v", z, z2)
+}
+
+func TestStressTest407(t *testing.T) {
+	for i := 0; i < 1024; i++ {
+		TestIssue_407(t)
+		if t.Failed() {
+			t.Errorf("Failed at iteration %d", i)
+			break
+		}
+	}
 }

--- a/encoding/dot/issues_test.go
+++ b/encoding/dot/issues_test.go
@@ -1,0 +1,70 @@
+package dot
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"gorgonia.org/gorgonia"
+)
+
+// unmangleName replaces the pointer-based name with the node name.
+//
+// Node names are unique in Gorgonia graphs.
+// However we cannot use node names in a `.dot` file because there are some names
+// that will not be accepted by graphviz.
+//
+// Hence we use pointers as part of the name of the nodes. A node will have a name
+// like "Node_0xbadb100d" (shout out to Tay Tay!) in the `.dot` file.
+//
+// unmangleName replaces `Node_0xbadc0ffee` with the name of the node for the purposes of testing.
+func unmangleName(marshalled string, node *gorgonia.Node) string {
+	name := node.Name()
+	ptrName := fmt.Sprintf("%p", node)
+
+	return strings.Replace(marshalled, ptrName, name, -1)
+}
+
+func TestIssue_407(t *testing.T) {
+	// originally written by @wzzhu
+
+	//assert := assert.New(t)
+	g := gorgonia.NewGraph()
+	var x, y, z *gorgonia.Node
+	var err error
+
+	// define the expression
+	x = gorgonia.NewScalar(g, gorgonia.Float64, gorgonia.WithName("x"))
+	y = gorgonia.NewScalar(g, gorgonia.Float64, gorgonia.WithName("y"))
+	if z, err = gorgonia.Add(x, y); err != nil {
+		t.Fatal(err)
+	}
+	b, err := Marshal(g)
+	if err != nil {
+		t.Errorf("Error printing graph: %v", err)
+	}
+	fst := fmt.Sprintf("\n%s", string(b))
+	fst = unmangleName(fst, x)
+	fst = unmangleName(fst, y)
+	fst = unmangleName(fst, z)
+
+	var x2, y2, z2 *gorgonia.Node
+
+	g2 := gorgonia.NewGraph()
+	x2 = gorgonia.NewScalar(g2, gorgonia.Float64, gorgonia.WithName("x"))
+	y2 = gorgonia.NewScalar(g2, gorgonia.Float64, gorgonia.WithName("y"))
+	if z2, err = gorgonia.Add(x2, y2); err != nil {
+		t.Fatal(err)
+	}
+	b2, err := Marshal(g2)
+	if err != nil {
+		t.Errorf("Error printing graph: %v", err)
+	}
+	snd := fmt.Sprintf("\n%s", string(b2))
+	snd = unmangleName(snd, x2)
+	snd = unmangleName(snd, y2)
+	snd = unmangleName(snd, z2)
+	//assert.Equal(fst, snd, "XXX")
+	// they are the same but due to Gonum's marshaling issue, it is marshalled undeterministically using a map.
+	t.Logf("%v %v", z, z2)
+}

--- a/encoding/dot/topgraph.go
+++ b/encoding/dot/topgraph.go
@@ -1,6 +1,8 @@
 package dot
 
 import (
+	"sort"
+
 	"gonum.org/v1/gonum/graph"
 	"gonum.org/v1/gonum/graph/encoding"
 	gonumDot "gonum.org/v1/gonum/graph/encoding/dot"
@@ -9,8 +11,8 @@ import (
 	internalEncoding "gorgonia.org/gorgonia/internal/encoding"
 )
 
-var (
-	subGraphs = map[internalEncoding.Group]subgrapher{
+func subGraphs() map[internalEncoding.Group]subgrapher {
+	return map[internalEncoding.Group]subgrapher{
 		internalEncoding.ConstantCluster: constantSubGraph{
 			DirectedBuilder: simple.NewDirectedGraph(),
 			name:            "Constants",
@@ -26,26 +28,36 @@ var (
 		},
 		internalEncoding.UndefinedCluster: exprSubGraph{
 			DirectedBuilder: simple.NewDirectedGraph(),
-			name:            "ExprGraph",
+			name:            "Undefined",
 		},
 	}
-)
+
+}
 
 type attributer []encoding.Attribute
 
 func (a attributer) Attributes() []encoding.Attribute { return a }
 
+func sortedKeys(m map[internalEncoding.Group]subgrapher) (retVal internalEncoding.Groups) {
+	for k := range m {
+		retVal = append(retVal, k)
+	}
+	sort.Sort(retVal)
+	return retVal
+}
+
 func generateDotGraph(g *gorgonia.ExprGraph) (graph.Graph, error) {
 	dg := simple.NewDirectedGraph()
 	copyGraph(dg, g)
 	nodes := dg.Nodes()
+	subgraphs := subGraphs()
 
 	for nodes.Next() {
 		n := nodes.Node()
 		if _, ok := n.(internalEncoding.Grouper); ok {
 			groups := n.(internalEncoding.Grouper).Groups()
 			for _, group := range groups {
-				if subgrapher, ok := subGraphs[group]; ok {
+				if subgrapher, ok := subgraphs[group]; ok {
 					subgrapher.(graph.DirectedBuilder).AddNode(n)
 				} else {
 					// check if we are in the ExprGraphCluster
@@ -56,7 +68,7 @@ func generateDotGraph(g *gorgonia.ExprGraph) (graph.Graph, error) {
 						name:            group.Name,
 					}
 					if groups.Have(internalEncoding.ExprGraphCluster) {
-						exprSubg := subGraphs[internalEncoding.ExprGraphCluster].(exprSubGraph)
+						exprSubg := subgraphs[internalEncoding.ExprGraphCluster].(exprSubGraph)
 						var ok bool
 						if _, ok = exprSubg.subs[group]; ok {
 							subgraph = exprSubg.subs[group]
@@ -67,14 +79,15 @@ func generateDotGraph(g *gorgonia.ExprGraph) (graph.Graph, error) {
 						continue
 					}
 					subgraph.AddNode(n)
-					subGraphs[group] = subgraph
+					subgraphs[group] = subgraph
 				}
 			}
 		}
 	}
-	subs := make([]gonumDot.Graph, 0, len(subGraphs))
-	for _, g := range subGraphs {
-		subs = append(subs, g)
+	subs := make([]gonumDot.Graph, 0, len(subgraphs))
+	keys := sortedKeys(subgraphs)
+	for _, k := range keys {
+		subs = append(subs, subgraphs[k])
 	}
 	return dotGraph{
 		Directed: dg,

--- a/internal/encoding/groups.go
+++ b/internal/encoding/groups.go
@@ -53,7 +53,7 @@ func (g Groups) Have(grp Group) bool {
 func (g Groups) Len() int { return len(g) }
 
 // Less checks if an ID is less than or not
-func (g Groups) Less(i, j int) bool { return g[i].ID < g[i].ID }
+func (g Groups) Less(i, j int) bool { return g[i].ID < g[j].ID }
 
 // Swap swaps the elements
 func (g Groups) Swap(i, j int) { g[i], g[j] = g[j], g[i] }

--- a/internal/encoding/groups.go
+++ b/internal/encoding/groups.go
@@ -28,21 +28,41 @@ type Grouper interface {
 type Groups []Group
 
 // Upsert the GroupID in the groups
-func (g *Groups) Upsert(grp Group) {
-	for i := 0; i < len(*g); i++ {
-		if (*g)[i].ID == grp.ID {
-			return
+func (g Groups) Upsert(grp Group) Groups {
+	for i := 0; i < len(g); i++ {
+		if (g)[i].ID == grp.ID {
+			return g
 		}
 	}
-	*g = append(*g, grp)
+	return append(g, grp)
 }
 
 // Have returns true if GroupID is in groups
-func (g *Groups) Have(grp Group) bool {
-	for i := 0; i < len(*g); i++ {
-		if (*g)[i].ID == grp.ID {
+func (g Groups) Have(grp Group) bool {
+	for i := 0; i < len(g); i++ {
+		if (g)[i].ID == grp.ID {
 			return true
 		}
 	}
 	return false
 }
+
+/* Groups by default sort by the group ID */
+
+// Len returns the length of a bag of groups
+func (g Groups) Len() int { return len(g) }
+
+// Less checks if an ID is less than or not
+func (g Groups) Less(i, j int) bool { return g[i].ID < g[i].ID }
+
+// Swap swaps the elements
+func (g Groups) Swap(i, j int) { g[i], g[j] = g[j], g[i] }
+
+// ByName is a sorting for a slice of groups, where the groups are sorted by name
+type ByName []Group
+
+func (g ByName) Len() int { return len(g) }
+
+func (g ByName) Less(i, j int) bool { return g[i].Name < g[j].Name }
+
+func (g ByName) Swap(i, j int) { g[i], g[j] = g[j], g[i] }

--- a/internal/encoding/groups_test.go
+++ b/internal/encoding/groups_test.go
@@ -1,6 +1,9 @@
 package encoding
 
-import "testing"
+import (
+	"sort"
+	"testing"
+)
 
 func TestGroups(t *testing.T) {
 	var g Groups
@@ -34,5 +37,41 @@ func TestGroups(t *testing.T) {
 }
 
 func TestSorts(t *testing.T) {
+	var gs Groups
+	g1 := NewGroup("A")
+	g2 := NewGroup("B")
+
+	gs = gs.Upsert(g2)
+	gs = gs.Upsert(g1)
+
+	t.Logf("%v", gs)
+	sort.Sort(gs)
+	if gs[0] != g1 {
+		t.Errorf("Groups: Expected the first element to be Group A. g1's ID %d", g1.ID)
+	}
+	if gs[1] != g2 {
+		t.Errorf("Groups: Expected the second element to be Group B, g2's ID %d", g2.ID)
+	}
+
+	t.Logf("%v", gs)
+
+	// reset
+
+	gs = gs[:0]
+	gs = gs.Upsert(g2)
+	gs = gs.Upsert(g1)
+
+	// check that resets work
+	if gs[0] != g2 {
+		t.Fatal("Must not proceed. Reset failed")
+	}
+
+	sort.Sort(ByName(gs))
+	if gs[0] != g1 {
+		t.Errorf("ByName: Expected the first element to be Group A. g1's name: %q", g1.Name)
+	}
+	if gs[1] != g2 {
+		t.Errorf("ByName: Expected the second element to be Group B. g2's name: %q", g2.Name)
+	}
 
 }

--- a/internal/encoding/groups_test.go
+++ b/internal/encoding/groups_test.go
@@ -6,15 +6,15 @@ func TestGroups(t *testing.T) {
 	var g Groups
 	g1 := NewGroup("g1")
 	g2 := NewGroup("g2")
-	g.Upsert(g1)
+	g = g.Upsert(g1)
 	if len(g) != 1 {
 		t.Fail()
 	}
-	g.Upsert(g1)
+	g = g.Upsert(g1)
 	if len(g) != 1 {
 		t.Fail()
 	}
-	g.Upsert(g2)
+	g = g.Upsert(g2)
 	if len(g) != 2 {
 		t.Fail()
 	}
@@ -30,5 +30,9 @@ func TestGroups(t *testing.T) {
 	if g.Have(NewGroup("g2")) {
 		t.Fail()
 	}
+
+}
+
+func TestSorts(t *testing.T) {
 
 }

--- a/nn.go
+++ b/nn.go
@@ -184,7 +184,7 @@ func Rectify(x *Node) (retVal *Node, err error) {
 	if retVal, err = ApplyOp(cmp, x, zero); err != nil {
 		return nil, errors.Wrap(err, applyOpFail)
 	}
-	retVal.groups.Upsert(group)
+	retVal.groups = retVal.groups.Upsert(group)
 
 	return HadamardProd(x, retVal)
 }
@@ -267,7 +267,7 @@ func Conv2d(im, filter *Node, kernelShape tensor.Shape, pad, stride, dilation []
 	if colIm, err = Im2Col(im, kernelShape, pad, stride, dilation); err != nil {
 		return
 	}
-	colIm.groups.Upsert(group)
+	colIm.groups = colIm.groups.Upsert(group)
 
 	layer := filter.Shape()[0]
 	kernel := filter.Shape()[1]
@@ -278,7 +278,7 @@ func Conv2d(im, filter *Node, kernelShape tensor.Shape, pad, stride, dilation []
 	if flattened, err = Reshape(filter, tensor.Shape{layer, kernel * row * col}); err != nil {
 		return
 	}
-	flattened.groups.Upsert(group)
+	flattened.groups = flattened.groups.Upsert(group)
 
 	// extract patch
 	batch := colIm.Shape()[0]
@@ -290,7 +290,7 @@ func Conv2d(im, filter *Node, kernelShape tensor.Shape, pad, stride, dilation []
 	if patch, err = Reshape(colIm, tensor.Shape{batch * m * n, z}); err != nil {
 		return
 	}
-	patch.groups.Upsert(group)
+	patch.groups = patch.groups.Upsert(group)
 
 	op := linAlgBinOp{
 		āBinaryOperator: matMulOperator,
@@ -301,16 +301,16 @@ func Conv2d(im, filter *Node, kernelShape tensor.Shape, pad, stride, dilation []
 	if colImLayer, err = ApplyOp(op, patch, flattened); err != nil {
 		return
 	}
-	colImLayer.groups.Upsert(group)
+	colImLayer.groups = colImLayer.groups.Upsert(group)
 
 	// now reshape and transpose the values back into the original order
 	var res *Node
 	if res, err = Reshape(colImLayer, tensor.Shape{batch, m, n, layer}); err != nil {
 		return
 	}
-	res.groups.Upsert(group)
+	res.groups = res.groups.Upsert(group)
 	ret, err := Transpose(res, 0, 3, 1, 2)
-	ret.groups.Upsert(group)
+	ret.groups = ret.groups.Upsert(group)
 	return ret, err
 }
 
@@ -356,7 +356,7 @@ func MaxPool2D(x *Node, kernel tensor.Shape, pad, stride []int) (*Node, error) {
 
 	op := newMaxPoolOp(xShape, kernel, pad, stride)
 	retVal, err := ApplyOp(op, x)
-	retVal.groups.Upsert(group)
+	retVal.groups = retVal.groups.Upsert(group)
 	return retVal, err
 }
 

--- a/node.go
+++ b/node.go
@@ -229,7 +229,7 @@ func WithGroupName(name string) NodeConsOpt {
 // withGroup is a node construction option to group a *Node within a particular group. This option is useful for debugging with graphs.
 func withGroup(group encoding.Group) NodeConsOpt {
 	f := func(n *Node) {
-		n.groups.Upsert(group)
+		n.groups = n.groups.Upsert(group)
 	}
 	return f
 }
@@ -245,11 +245,11 @@ func (n *Node) Groups() encoding.Groups {
 
 	switch {
 	case isConst:
-		n.groups.Upsert(encoding.ConstantCluster)
+		n.groups = n.groups.Upsert(encoding.ConstantCluster)
 	case isInput:
-		n.groups.Upsert(encoding.InputCluster)
+		n.groups = n.groups.Upsert(encoding.InputCluster)
 	default:
-		n.groups.Upsert(encoding.ExprGraphCluster)
+		n.groups = n.groups.Upsert(encoding.ExprGraphCluster)
 	}
 	return n.groups
 }

